First pass at list page

taysea opened the pull request:
[Preview](https://deploy-preview-956--keen-mayer-a86c8b.netlify.app/templates/lists)

First pass at list page. Includes:
- [x] Guidance, from [first pass by Maggie](https://github.com/hpe-design/design-system/issues/376#issuecomment-626829288)
- [x] Variants, matching formats from [Figma](https://www.figma.com/file/tMYTr17r6hhGajSlNRiPje/HPE-List-Component?node-id=19%3A135) but with some mock data
- [x] Theme changes local, but PR will be opened on grommet-theme-hpe

To-do:
- [x] NO LONGER NECESSARY: Add hover theming to list item in grommet (currently hover is only possible when list item has an onClick, however we may not want the entire list item to have onClick if the list item contains a nested action button). This functionality will be added when this grommet PR merges: https://github.com/grommet/grommet/pull/4279
EDIT: After a design discussion, hover should only appear if the list item is clickable. No further dev work required here.
- [x] Need to discuss what happens when List item is intended to be clickable and there is an actions button. Is this a case?
--> EDIT: if list item is clickable, then there should be no action button. this needs to be documented in guidance
Closes #941

halocline commented on aries-site/src/data/structures/templates.js:
```diff
@@ -14,9 +19,14 @@ export const templates = [
         fit: 'contain',
       },
     },
+    relatedContent: ['Dashboards', 'Page Layouts'],
```
Do we want to add in 'Cards' as well? Will be 'Coming Soon' content immediately, but should be filled in in the next week or so.

taysea replied:
Yes, let's add Cards.

halocline commented on aries-site/src/examples/templates/list-views/ListIconIdentifierExample.js:
```diff
@@ -0,0 +1,64 @@
+  },
+];
+
+export const ListIconIdentifierExample = () => {
```
I am seeing the UserAgent focus color come through. Is that what we want?
![Screen Shot 2020-07-06 at 1 28 28 PM](https://user-images.githubusercontent.com/1756948/86632532-c0d91400-bf8c-11ea-9cf2-60b3c94210cc.png)

taysea replied:
I am not sure we want this. It is occurring in the generic grommet storybook as well, so may be result of implementation and a fix that would happen on the grommet side? https://storybook.grommet.io/?path=/story/list--onclickitem

halocline replied:
Agree. We should create an issue in Design System for prioritization and tracking purposes and link it to a corresponding issue in Grommet.

taysea replied:
Great, I will open both of those tickets.

taysea replied:
Being tracked on Design System: https://github.com/hpe-design/design-system/issues/965
and on grommet: https://github.com/grommet/grommet/issues/4278

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+  figma="https://www.figma.com/file/tMYTr17r6hhGajSlNRiPje/HPE-List-Component?node-id=19%3A135"
+  width="100%"
+>
+  <ListImageIdentifierExample />
```
Fullscreen mode width seems to be too narrow:
![Screen Shot 2020-07-06 at 1 30 58 PM](https://user-images.githubusercontent.com/1756948/86632858-27f6c880-bf8d-11ea-9fa7-d72de1529086.png)

halocline replied:
I would think including the Mobile view would be beneficial. Thoughts?

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+
+Single-line list with an image, primary title, short status tag, and an action button.
+
+<Example
```
Fullscreen mode width seems to be too narrow:
![Screen Shot 2020-07-06 at 1 30 58 PM](https://user-images.githubusercontent.com/1756948/86632858-27f6c880-bf8d-11ea-9fa7-d72de1529086.png)

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+
+Single-line list with an image, primary title, short status tag, and an action button.
+
+<Example
```
I would think including the Mobile view would be beneficial. Thoughts?

taysea replied:
Yes, let's add mobile view. I am working to expand the functionality of `Example` to better handle showing responsive views.

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+
+Two-line list with an image, primary title, description, and an action button.
+
+<Example
```
Do we provide any guidance as to a max-width for a List? On widescreens it becomes difficult to track from left to righthand side:
![Screen Shot 2020-07-06 at 1 38 59 PM](https://user-images.githubusercontent.com/1756948/86633643-27126680-bf8e-11ea-94b1-da89b6ca32b0.png)

halocline replied:
At one point, I thought we had list examples with a bottom border. Did that get eliminated?

taysea replied:
That's a great point. We should have some guidance on this.

halocline replied:
> Also, regarding bottom border, the Figma designs now call for now border. However, I will reach out to see if we want to demonstrate any cases with a border and look into implementation. Since the border removal will be handled in the theme, it seems like we'd want border to be accessible via props to avoid use of Theme.Extend

Clarifying:
- We will handle the most common execution of border (no border) in the theme.
- If a border is desired, we will use [List's border prop](https://v2.grommet.io/list#border)

taysea replied:
We will want to add guidance on when List with border would be appropriate as well. I'm not sure on the guidance for when to use border or not, so I will reach out to the design team.

cherih replied:
Need to change the primary example on this page. It uses outdated styling. Pick one of the other examples to show instead.

BTW - I filed #974 - Update Figma Designs for Lists tesmplate - to get some help from either Greg or Vicky on updates to the Figma designs for this.

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
 
 <Example
   code="https://raw.githubusercontent.com/hpe-design/design-system/master/aries-site/src/examples/templates/list-views/ListScreenExample.js"
+  docs="https://v2.grommet.io/list?theme=hpe#props"
```
Should we have a Figma link to this thumbnail? https://www.figma.com/file/tMYTr17r6hhGajSlNRiPje/HPE-List-Component?node-id=36%3A19

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+<Example
+  code="https://raw.githubusercontent.com/hpe-design/design-system/master/aries-site/src/examples/templates/list-views/ListImageIdentifierExample.js"
+  docs="https://v2.grommet.io/list?theme=hpe#props"
+  figma="https://www.figma.com/file/tMYTr17r6hhGajSlNRiPje/HPE-List-Component?node-id=19%3A135"
```
Have we decided to include the same link to Figma files in all examples now? This seems to be a departure from what we were doing with Components.

taysea replied:
You're right. I think just the first one should have a figma link.

halocline commented on aries-site/src/examples/templates/list-views/ListImageIdentifierExample.js:
```diff
@@ -0,0 +1,58 @@
+        <Box direction="row" gap="small" key={index}>
+          <Box width="xxsmall" height="xxsmall">
+            <Image
+              src="https://images.unsplash.com/photo-1487088678257-3a541e6e3922?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2167&q=80"
```
Should the image url be part of the data object? Perhaps tied to the location?

If lists are about scan-ability, then the inclusion of an image should assist the scan-ability. If we display the same image for all, then it is not adding any value and should be removed.

taysea replied:
Agreed, will put images tied to location.

halocline commented on aries-site/src/pages/templates/lists.mdx:
```diff
@@ -1,16 +1,109 @@
+
+### Image + Name + Description + Action
+
+Two-line list with an image, primary title, description, and an action button.
```
In each of these variants, I feel like we should be elaborating on the unique element being added to the example. In this case, the image is what is being added, so we should explain how it should be used to assist scan-ability -- specific to this example, the image should be tied to either the User or the Location. If for the User, each image would be unique and assist the end-user in making a deeper cognitive connection with each User in the list; if for the location, then allows the end-user to infer common associations / mental groupings between individual users.

halocline replied:
Basically, this relates to discussion about "we are doing a good job with the 'What', but need to get better in our guidance about filling in the 'Why', 'How', and 'When' to use.

taysea replied:
I agree completely. I put this in as starter text, but was thinking Cheri would be able to add in more details

halocline replied:
Just put a summary of my thoughts in a review comment. Code-wise things look good. Main reactions are to the content and story we are telling. Don't know the best way to proceed. Perhaps, do like Page Layouts --> publish --> demo and/or summarize with callouts for areas needing attention / thought starters -->  get folks to react and contribute their thoughts ???

cherih replied:
Agree with Matt's comment about ordering the examples from simple to complex, allowing you to build on the previous ones.

Would like to see examples of other screen sizes too, especially mobile.

Feel free to add a github issue for me to help out with this, with what is most needed.

taysea replied:
@cherih I've created an issue here summarizing some of the content additions we'd like to add to this page: https://github.com/hpe-design/design-system/issues/964

cherih replied:
@taysea - thanks I'll take a look at #964.

cherih replied:
Just thinking about the Figma design examples some more and the previous comments about ordering examples.

Is it possible to start with the simplest example of all (i.e. Name only)? 
Then create the more complex examples step by step.

Also, maybe include fewer examples with the Actions menu in the list?
We are advising against using it for GLC. It was confusing to users in usability tests and it doesn't support multi-select.

halocline commented on aries-site/src/examples/templates/list-views/ListNameOptionActionExample.js:
```diff
@@ -0,0 +1,51 @@
+      data={data}
+      action={item => (
+        <Box direction="row" align="center" gap="medium">
+          <Text>{item.option}</Text>
```
These don't show up in Figma, but wondering about the associating color indicator along side status:
![Screen Shot 2020-07-06 at 2 24 20 PM](https://user-images.githubusercontent.com/1756948/86638433-84a9b180-bf94-11ea-88ac-2ba2b595d618.png)

taysea replied:
I like this idea. I will adjust one example to show this.

cherih commented:
@taysea - thank you for adding the background to the examples. That really helps visually with the mobile examples.

ericsoderberghp commented:
Some feedback:

- Why are we repeating:

ericsoderberghp commented:
Some feedback ...

- Why are we repeating this twice on the page?:

> HPE Design System list view templates are go-to patterns for deploying homogeneous data such as services, devices, users, and more. List views are optimized for scanability and reading comprehension. Each list item provides users focused information and identity labels to aid selection, decision making, and action.

- I wouldn't think of a list as something used for "deploying homogeneous data". Perhaps "displaying"?
- I think we should be more consistent about terms like "homogeneous" or "related". I think this is a good topic, what criteria should be used for placing things in a list. For instance, why not put everything in a single global list with searching and filtering and call it done?
- I wouldn't have thought of a list primarily as "an effective way to display dense content to users." I think a list is primarily used for: A) looking for something, coupled with search or filters, provides a quick means of scanning to find something, B) aggregating, coupled with search or filters, provides a summary of what items match the search/filter criteria and verification of the matching items, C) selecting multiple items to perform an action with them.
- I think we should add a section for "Navigation" and mention that clicking an item in a list can be a means of navigating to another page/screen or as a means to display further detail for that item in either a layer or separate panel.
- The section on "Usage" seems primarily to be about "types", which don't quite seem to be the same thing.
- The Usage section mentions "Grid lists" but there no other references to that and I'm not sure what it refers to.
- I would put the List vs. Table section last. Where it is now it kind of breaks up the flow.
- I'm not sure how selection relates to action in:

> If items can be selected using only the List contents, then use an Action Menu in a List item. Otherwise, go to a detail view to see additional information for the List item and use an Action Menu there.

Do we mean to say that if a List item has a context for more details, put any actions there?

- In:

> For really long lists (i.e. hundreds or thousands), consider limiting the number of items that can be shown at a time and provide some mechanism for accessing the rest of the items

How about adding something like "such as infinite scrolling with lazy retrieval and rendering"?

- In:

> Avoid paging, if possible. It is difficult to do and can result in a confusing user experience

"difficult to do" seems a bit vague. How about "Avoid pagination. It tends to cause a more complicated user experience, especially when selection and/or automatic updating are involved."

- In:

> Lists should span the width of the page it is in, unless it is placed in a constraining box.

How about "Lists should span the width of the container they are laid out in"

- Regarding accessibility, I think we should discuss keyboard interaction and screen reader behavior. For instance, when list items are interactive, the focus is always on the list itself and up/down keys are used to move an indicator indicating which item will be used when the Enter key is pressed. Each list should also have a clear `aria-label` indicating what the list contains and a summary of the contents, such as how many items are in the list.

- For the "Name" variation: https://deploy-preview-956--keen-mayer-a86c8b.netlify.app/templates/lists#name. I'm not sure when I would use this. I think we need some text explaining when it would be used. I also wonder if we should omit it as a variation, or not make it the first one, since it doesn't seem like the typical variation.

- There is no "Action" in https://deploy-preview-956--keen-mayer-a86c8b.netlify.app/templates/lists#icon-name-option-action

- I think the examples should all have more real world data in them. "Title 1", "Server 1", "User 1" seem too generic. The IP address, state, and action menu items feel real.

- I don't think we'd recommend the "Location: " prefix in https://deploy-preview-956--keen-mayer-a86c8b.netlify.app/templates/lists#image-name-description. We'd likely just have "Houston TX"

taysea commented:
Thanks for all of this feedback, @ericsoderberghp. I've incorporated the wording suggestions into the sections and also modified the Examples to include more real world feeling data.

Regarding the "Action Menu" feedback, I will reach out to @cherih to clarify the wording there.

ericsoderberghp commented:
I'm OK with merging this now. Thanks for the updates.

cherih commented:
@ericsoderberghp - thank you for all of the feedback!

@taysea - thank you for rolling in Eric's feedback. I'll look at the Action Menu comments today or tomorrow, and work with you on updating them.
